Content.Server/Ninja/Systems/NinjaSystem.cs

```diff
@@ -0,0 +1,678 @@
+        {
+            if (_tags.HasTag(target, comp.EmagImmuneTag))
+                return;
+
+            var handled = _emag.DoEmagEffect(user, target);
+            if (!handled)
+                return;
+
+            _popups.PopupEntity(Loc.GetString("emag-success", ("target", Identity.Entity(target, EntityManager))), user,
+                user, PopupType.Medium);
+            _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(user):player} doorjacked {ToPrettyString(target):target}");
+
+            args.Handled = true;
+            return;
+        }
+
+        // stun ability
+        if (user != target && HasComp<StaminaComponent>(target))
+        {
+            // take charge from battery
+            if (!GetNinjaBattery(user, out var battery) || !battery.TryUseCharge(comp.StunCharge))
+            {
+                _popups.PopupEntity(Loc.GetString("ninja-no-power"), user, user);
+                return;
+            }
+
+            // not holding hands with target so insuls don't matter
+            args.Handled = _electrocution.TryDoElectrocution(target, uid, comp.StunDamage, comp.StunTime, false, ignoreInsulation: true);
+            return;
+        }
+
+        // drain ability
+        if (HasComp<PowerNetworkBatteryComponent>(target))
+        {
+            if (!HasComp<BatteryComponent>(target))
+                return;
+
+            comp.CancelToken = new CancellationTokenSource();
+            var doafterArgs = new DoAfterEventArgs(user, comp.DrainTime, comp.CancelToken.Token, used: uid)
+            {
+                BreakOnDamage = true,
+                BreakOnStun = true,
+                BreakOnUserMove = true,
+                MovementThreshold = 0.5f,
+                UsedCancelledEvent = new GloveActionCancelledEvent(),
+                UsedFinishedEvent = new DrainSuccessEvent(user, target)
+            };
+
+            _doafter.DoAfter(doafterArgs);
+            args.Handled = true;
+            return;
+        }
+
+        // download ability
+        if (TryComp<TechnologyDatabaseComponent>(target, out var database))
+        {
+            // fail fast if theres no tech right now
+            if (database.TechnologyIds.Count == 0)
+            {
+                _popups.PopupEntity(Loc.GetString("ninja-download-fail"), user, user);
+                return;
+            }
+
+            comp.CancelToken = new CancellationTokenSource();
+            var doafterArgs = new DoAfterEventArgs(user, comp.DownloadTime, comp.CancelToken.Token, used: uid)
+            {
+                BreakOnDamage = true,
+                BreakOnStun = true,
+                BreakOnUserMove = true,
+                MovementThreshold = 0.5f,
+                UsedCancelledEvent = new GloveActionCancelledEvent(),
+                UsedFinishedEvent = new DownloadSuccessEvent(user, target)
+            };
+
+            _doafter.DoAfter(doafterArgs);
+            args.Handled = true;
+            return;
+        }
```

Comment:
you mean like having a single DoAfter call then making the ability things set the event or what

Reply:
What you have right now might cause multiple doafters to run on the gloves since gloves have Drain, Download and Terror. So what you can do is have the events set up like this: 

`SubscribeLocalEvent<NinjaTerrorComponent, DoAfterEvent<DataHere>>(OnTerrorDoAfter);`

`<DataHere>` can be any kind of struct or class to have a distinction between the doafters.

Reply:
how i have it right now works since they all centrally check if the gloves have an active doafter, is it still worth using unique data to stop them running in the first place?